R/permutation.R

```diff
@@ -0,0 +1,703 @@
+        alpha = k/( n-1 )    # Method B
+        beta = 1 - 2*alpha
+        q = sort( placebo_estimates )[ c(1+k, n - k) ]
+    } else {
+        alpha = (1 - beta)/2
+        k = alpha * (n-1)  # QUESTION: Why n-1 here????
+        if ( round_beta ) {
+            k = pmax( 1, round( k ) )
+            alpha = k/(n-1)
+            beta = 1 - 2*alpha
+            q = sort( placebo_estimates )[ c(1+k, n - k) ]
+        } else {
+            q = quantile( placebo_estimates, probs = c( alpha, 1 - alpha ) )
+        }
+    }
+
+    del = as.numeric( diff( q ) )
+    z = -qnorm( alpha )
+    SE_imp = del / (2*z)
+
+    res = data.frame( beta = beta, k = k,
+                      q_low = q[[1]], q_high = q[[2]],
+                      range = del, z = z, SE_imp = SE_imp )
+    res
+}
+
+
+#' Construct an organized dataframe with outcome data in long format
+#' from augsynth object
+#'
+#' @noRd
+get_long_data <- function( augsynth ) {
+
+    wide_data <- augsynth$data
+
+    tx_id <- augsynth$trt_unit
+    control_ids = which( wide_data$trt == 0 )
+
+    all_ids = 1:nrow(wide_data$X)
+    all_ids[ control_ids ] = rownames( augsynth$weights )
+    all_ids[ which( wide_data$trt == 1 ) ] = tx_id
+
+    df <- bind_cols(wide_data$X, wide_data$y) %>%
+        mutate(!!as.name(augsynth$unit_var) := all_ids) %>%
+        pivot_longer(!augsynth$unit_var,
+                     names_to = augsynth$time_var,
+                     values_to = 'Yobs',
+        ) %>%
+        mutate(!!as.name(augsynth$time_var) := as.numeric(!!as.name(augsynth$time_var)),
+               ever_Tx = ifelse(!!as.name(augsynth$unit_var) == augsynth$trt_unit, 1, 0))
+
+    df
+}
+
+
+
+add_placebo_distribution <- function(augsynth) {
+
+    # Run permutations
+    ests <- get_placebo_gaps(augsynth, att = FALSE)
+    time_cols = 2:ncol(ests)
+
+    ests$trt = augsynth$data$trt
+
+    lest = ests %>%
+        pivot_longer( cols = all_of( time_cols ),
+                      names_to = augsynth$time_var, values_to = "Yhat" ) %>%
+        mutate(!!as.name(augsynth$time_var) := as.numeric(!!as.name(augsynth$time_var)))
+
+    df <- get_long_data( augsynth )
+
+    ##### Make dataset of donor units with their weights #####
+    units = dplyr::select( df, !!as.name(augsynth$unit_var), ever_Tx ) %>%
+        unique()
+
+    tx_col = augsynth$unit_var
+    weights = data.frame( tx_col = rownames( augsynth$weights ),
+                          weights = augsynth$weights,
+                          stringsAsFactors = FALSE)
+    colnames(weights)[1] <- augsynth$unit_var
+
+    units = merge( units, weights, by=augsynth$unit_var, all=TRUE )
+
+    # Zero weights for tx units.
+    units$weights[ units$ever_Tx == 1 ] = 0
+
+    # confirm that we have placebos for every entity over every observed time period
+    stopifnot("Placebos do not cover every entity over every observed time period." = (ncol(augsynth$data$X) + ncol(augsynth$data$y)) * nrow(augsynth$data$y) == nrow(lest))
```

Comment:
Same as above: have you run into a situation where this check fails, is there a minimum reproducible example, and is there a way to fix the code so it doesn't happen or is this due to users putting in something incorrectly?